lib/internal/fs/dir.js

```diff
@@ -0,0 +1,182 @@
+    ERR_MISSING_ARGS
+  }
+} = require('internal/errors');
+
+const { FSReqCallback } = binding;
+const internalUtil = require('internal/util');
+const {
+  getDirent,
+  getOptions,
+  getValidatedPath,
+  handleErrorFromBinding
+} = require('internal/fs/utils');
+
+const kDirHandle = Symbol('kDirHandle');
+const kDirPath = Symbol('kDirPath');
+const kDirOptions = Symbol('kDirOptions');
+const kDirReadPromisified = Symbol('kDirReadPromisified');
+const kDirClosePromisified = Symbol('kDirClosePromisified');
+
+class Dir {
+  constructor(handle, path, options) {
+    if (handle == null) throw new ERR_MISSING_ARGS('handle');
+    this[kDirHandle] = handle;
+    this[kDirPath] = path;
+
+    this[kDirOptions] = getOptions(options, {
+      encoding: 'utf8'
+    });
+
+    this[kDirReadPromisified] = internalUtil.promisify(this.read).bind(this);
+    this[kDirClosePromisified] = internalUtil.promisify(this.close).bind(this);
+  }
+
+  // TODO(Fishrock123): make this return absolute path
+  get path() {
+    return pathModule.toNamespacedPath(this[kDirPath]);
+  }
+
+  read(options, callback) {
+    callback = typeof options === 'function' ? options : callback;
+    if (callback === undefined) {
+      return this[kDirReadPromisified](options);
+    } else if (typeof callback !== 'function') {
+      throw new ERR_INVALID_CALLBACK(callback);
+    }
+    options = getOptions(options, this[kDirOptions]);
+
+    const req = new FSReqCallback();
+    req.oncomplete = (err, result) => {
+      if (err || result === null) {
+        return callback(err, result);
+      }
+      getDirent(this[kDirPath], result[0], result[1], callback);
+    };
+
+    this[kDirHandle].read(
+      options.encoding,
+      req
+    );
+  }
+
+  readSync(options) {
+    options = getOptions(options, this[kDirOptions]);
+
+    const ctx = { path: this[kDirPath] };
+    const result = this[kDirHandle].read(
+      options.encoding,
+      undefined,
+      ctx
+    );
+    handleErrorFromBinding(ctx);
+
+    if (result === null) {
+      return result;
+    }
+
+    return getDirent(this[kDirPath], result[0], result[1]);
+  }
+
+  close(callback) {
+    if (callback === undefined) {
+      return this[kDirClosePromisified]();
+    } else if (typeof callback !== 'function') {
+      throw new ERR_INVALID_CALLBACK(callback);
+    }
+    const req = new FSReqCallback();
+    req.oncomplete = callback;
+    this[kDirHandle].close(req);
+  }
+
+  closeSync() {
+    const ctx = { path: this[kDirPath] };
+    const result = this[kDirHandle].close(undefined, ctx);
+    handleErrorFromBinding(ctx);
+    return result;
+  }
+
+  async* entries() {
+    try {
+      while (true) {
+        const result = await this[kDirReadPromisified]();
+        if (result === null) {
+          break;
+        }
+        yield result;
+      }
+    } finally {
+      // TODO(Fishrock123): Write our own iterable to overcome the limitations
+      // of generator functions.
+      this.closeSync();
```

Comment:
I think this should be addressed before this PR lands.

That being said, I’d be curious how exactly these limitations look like, because that would seem like an issue with the language that might need fixing.